arangod/Cluster/SynchronizeShard.cpp

```diff
@@ -815,157 +819,248 @@ bool SynchronizeShard::first() {
 
       auto details = std::make_shared<VPackBuilder>();
 
-      Result syncRes = replicationSynchronize(
+      res = replicationSynchronize(
         collection, config.slice(), APPLIER_DATABASE, details);
 
       auto sy = details->slice();
       auto const endTime = system_clock::now();
-      bool longSync = false;
-
       // Long shard sync initialisation
       if (endTime-startTime > seconds(5)) {
-        LOG_TOPIC(WARN, Logger::MAINTENANCE)
+        LOG_TOPIC(INFO, Logger::MAINTENANCE)
           << "synchronizeOneShard: long call to syncCollection for shard"
-          << shard << " " << syncRes.errorMessage() <<  " start time: "
-          << timepointToString(startTime) <<  "end time: "
+          << shard << " " << res.errorMessage() <<  " start time: "
+          << timepointToString(startTime) <<  ", end time: "
           << timepointToString(system_clock::now());
-        longSync = true;
       }
 
-      //
-      if (!syncRes.ok()) {
+      // If this did not work, then we cannot go on:
+      if (!res.ok()) {
 
         std::stringstream error;
         error << "could not initially synchronize shard " << shard << ": "
-              << syncRes.errorMessage();
+              << res.errorMessage();
         LOG_TOPIC(ERR, Logger::MAINTENANCE) << "SynchronizeOneShard: " << error.str();
         _result.reset(TRI_ERROR_INTERNAL, error.str());
         return false;
 
-      } else {
-
-        VPackSlice collections = sy.get(COLLECTIONS);
-
-        if (collections.length() == 0 ||
-            collections[0].get("name").copyString() != shard) {
+      }
 
-          if (longSync) {
-            LOG_TOPIC(ERR, Logger::MAINTENANCE)
-              << "synchronizeOneShard: long sync, before cancelBarrier"
-              << timepointToString(system_clock::now());
-          }
-          cancelBarrier(ep, database, sy.get(BARRIER_ID).getNumber<int64_t>(), clientId);
-          if (longSync) {
-            LOG_TOPIC(ERR, Logger::MAINTENANCE)
-              << "synchronizeOneShard: long sync, after cancelBarrier"
-              << timepointToString(system_clock::now());
-          }
+      // From here on, we have to call `cancelBarrier` in case of errors
+      // as well as in the success case!
+      int64_t barrierId = sy.get(BARRIER_ID).getNumber<int64_t>();
+      TRI_DEFER(cancelBarrier(ep, database, barrierId, clientId));
```

Comment:
shouldn't you send cancelBarrier before logging you are done? I think you can use a scopeGuard and `guard.fire()` to trigger it manually

Reply:
In this case it is actually fine with DEFER.
We are in a try catch scope, this is either left by error or exception => cancelBarrier
Or it is left successfully => cancelBarrier before logging for the user.